controllers/clusteroperator.go

```diff
@@ -0,0 +1,152 @@
+// status condition change.
+type StatusReason string
+
+const (
+	clusterOperatorName = "baremetal"
+
+	// OperatorDisabled represents a Disabled ClusterStatusConditionTypes
+	OperatorDisabled osconfigv1.ClusterStatusConditionType = "Disabled"
+
+	// ReasonEmpty is an empty StatusReason
+	ReasonEmpty StatusReason = ""
+	// ReasonComplete is an empty StatusReason
+	ReasonComplete StatusReason = "DeployComplete"
+	// ReasonSyncing is an complete StatusReason
+	ReasonSyncing StatusReason = "SyncingResources"
+	// ReasonSyncFailed is an failed StatusReason
+	ReasonSyncFailed StatusReason = "SyncingFailed"
+	// ReasonUnsupported is an unsupported StatusReason
+	ReasonUnsupported StatusReason = "UnsupportedPlatform"
+)
+
+// defaultStatusConditions returns the default set of status conditions for the
+// ClusterOperator resource used on first creation of the ClusterOperator.
+func defaultStatusConditions() []osconfigv1.ClusterOperatorStatusCondition {
+	return []osconfigv1.ClusterOperatorStatusCondition{
+		setStatusCondition(osconfigv1.OperatorProgressing, osconfigv1.ConditionFalse, "", ""),
+		setStatusCondition(osconfigv1.OperatorDegraded, osconfigv1.ConditionFalse, "", ""),
+		setStatusCondition(osconfigv1.OperatorAvailable, osconfigv1.ConditionFalse, "", ""),
+		setStatusCondition(osconfigv1.OperatorUpgradeable, osconfigv1.ConditionTrue, "", ""),
+		setStatusCondition(OperatorDisabled, osconfigv1.ConditionFalse, "", ""),
+	}
+}
+
+// relatedObjects returns the current list of ObjectReference's for the
+// ClusterOperator objects's status.
+func relatedObjects() []osconfigv1.ObjectReference {
+	return []osconfigv1.ObjectReference{
+		{
+			Group:    "",
+			Resource: "namespaces",
+			Name:     ComponentNamespace,
+		},
+	}
+}
+
+// createClusterOperator creates the ClusterOperator and updates its status.
+func (r *ProvisioningReconciler) createClusterOperator() (*osconfigv1.ClusterOperator, error) {
+	defaultCO := &osconfigv1.ClusterOperator{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterOperator",
+			APIVersion: "config.openshift.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: clusterOperatorName,
+		},
+		Status: osconfigv1.ClusterOperatorStatus{
+			Conditions:     defaultStatusConditions(),
+			RelatedObjects: relatedObjects(),
+		},
+	}
+	operatorv1helpers.SetOperandVersion(&defaultCO.Status.Versions, osconfigv1.OperandVersion{Name: "operator", Version: os.Getenv("RELEASE_VERSION")})
+
+	co, err := r.OSClient.ConfigV1().ClusterOperators().Create(context.Background(), defaultCO, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	co.Status = defaultCO.Status
+	return r.OSClient.ConfigV1().ClusterOperators().UpdateStatus(context.Background(), co, metav1.UpdateOptions{})
+}
+
+// getOrCreateClusterOperator gets the existing CO, failing which it creates a new CO.
+func (r *ProvisioningReconciler) getOrCreateClusterOperator() (*osconfigv1.ClusterOperator, error) {
```

Comment:
Can we add a unit -test for the case where we are able to get an existing CO when it exists?

Reply:
Done. Thanks.